cloudinit/net/freebsd.py

```diff
@@ -0,0 +1,167 @@
+# This file is part of cloud-init. See LICENSE file for license information.
+
+import re
+
+from cloudinit import log as logging
+from cloudinit import net
+from cloudinit import util
+from cloudinit.distros import rhel_util
+from cloudinit.distros.parsers.resolv_conf import ResolvConf
+
+from . import renderer
+
+LOG = logging.getLogger(__name__)
+
+
+class Renderer(renderer.Renderer):
+    resolv_conf_fn = 'etc/resolv.conf'
+    rc_conf_fn = 'etc/rc.conf'
+
+    def __init__(self, config=None):
+        if not config:
+            config = {}
+        self.dhcp_interfaces = []
+        self._postcmds = config.get('postcmds', True)
+
+    def _write_ifconfig_entries(self, settings, target=None):
+        ifname_by_mac = net.get_interfaces_by_mac()
+        for interface in settings.iter_interfaces():
+            device_name = interface.get("name")
+            device_mac = interface.get("mac_address")
+            if device_name and re.match(r'^lo\d+$', device_name):
+                continue
+            if device_mac not in ifname_by_mac:
+                LOG.info('Cannot find any device with MAC %s', device_mac)
+            elif device_mac and device_name:
+                cur_name = ifname_by_mac[device_mac]
+                if cur_name != device_name:
+                    LOG.info('netif service will rename interface %s to %s',
+                             cur_name, device_name)
+                    rhel_util.update_sysconfig_file(
+                        util.target_path(target, self.rc_conf_fn), {
+                            'ifconfig_%s_name' % cur_name: device_name})
+            else:
+                device_name = ifname_by_mac[device_mac]
+
+            LOG.info('Configuring interface %s', device_name)
+            ifconfig = 'DHCP'  # default
+
+            for subnet in interface.get("subnets", []):
+                if ifconfig != 'DHCP':
+                    LOG.info('The FreeBSD provider only set the first subnet.')
+                    break
+                if subnet.get('type') == 'static':
+                    LOG.debug('Configuring dev %s with %s / %s', device_name,
+                              subnet.get('address'), subnet.get('netmask'))
+                # Configure an ipv4 address.
+                    ifconfig = (
+                            subnet.get('address') + ' netmask ' +
+                            subnet.get('netmask'))
```

Comment:
getting a stacktrace here:

```python-traceback
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/cloudinit/cmd/main.py", line 653, in status_wrapper
    ret = functor(name, args)
  File "/usr/local/lib/python3.6/site-packages/cloudinit/cmd/main.py", line 362, in main_init
    init.apply_network_config(bring_up=bool(mode != sources.DSMODE_LOCAL))
  File "/usr/local/lib/python3.6/site-packages/cloudinit/stages.py", line 710, in apply_network_config
    return self.distro.apply_network_config(netcfg, bring_up=bring_up)
  File "/usr/local/lib/python3.6/site-packages/cloudinit/distros/__init__.py", line 178, in apply_network_config
    dev_names = self._write_network_config(netconfig)
  File "/usr/local/lib/python3.6/site-packages/cloudinit/distros/freebsd.py", line 197, in _write_network_config
    return self._supported_write_network_config(netconfig)
  File "/usr/local/lib/python3.6/site-packages/cloudinit/distros/__init__.py", line 93, in _supported_write_network_config
    renderer.render_network_config(network_config)
  File "/usr/local/lib/python3.6/site-packages/cloudinit/net/renderer.py", line 56, in render_network_config
    templates=templates, target=target)
  File "/usr/local/lib/python3.6/site-packages/cloudinit/net/freebsd.py", line 145, in render_network_state
    self._write_network(network_state, target=target)
  File "/usr/local/lib/python3.6/site-packages/cloudinit/net/freebsd.py", line 138, in _write_network
    self._write_ifconfig_entries(settings, target=target)
  File "/usr/local/lib/python3.6/site-packages/cloudinit/net/freebsd.py", line 59, in _write_ifconfig_entries
    subnet.get('netmask'))
TypeError: must be str, not NoneType
```

Reply:
Addressed in https://github.com/goneri/cloud-init/commit/3e6a1199031d9477ddd8d33234023f942a2ba79f